src/Helper/Quote.php

```diff
@@ -108,17 +112,51 @@ public function __construct(
      */
     public function getValidatedCoupons()
     {
-        $coupons = [];
+        $coupons                     = [];
+        $returnInvalidCartRuleCoupon = false;
+        $discountAmount              = $this->quote->getSubtotal() - $this->quote->getSubtotalWithDiscount();
+        $couponsIncludeTax           = $this->taxHelper->couponInclTax();
+        $quoteCurrency               = $this->quote->getStoreCurrencyCode();
         foreach ($this->sgBase->getExternalCoupons() as $coupon) {
+            if ($coupon->getCode() === self::CART_RULE_COUPON_CODE) {
+                $returnInvalidCartRuleCoupon = true;
+                continue;
+            }
             if (!$coupon->getNotValidMessage()) {
-                $amount = $this->quote->getSubtotal() - $this->quote->getSubtotalWithDiscount();
                 $coupon->setIsValid(true);
-                $coupon->setCurrency($this->quote->getStoreCurrencyCode());
-                $coupon->setIsFreeShipping((bool) $this->quote->getShippingAddress()->getFreeShipping());
-                $this->taxHelper->couponInclTax() ? $coupon->setAmountGross($amount) : $coupon->setAmountNet($amount);
+                $coupon->setCurrency($quoteCurrency);
+                $coupon->setIsFreeShipping((bool)$this->quote->getShippingAddress()->getFreeShipping());
+                $couponsIncludeTax ? $coupon->setAmountGross($discountAmount) : $coupon->setAmountNet($discountAmount);
             }
             $coupons[] = $this->couponHelper->dataToEntity($coupon->toArray());
         }
+        if (empty($coupons) && !empty($discountAmount)) {
+            $couponData                  = [
+                'code'             => self::CART_RULE_COUPON_CODE,
+                'name'             => __(self::CART_RULE_COUPON_NAME),
+                'is_valid'         => true,
+                'currency'         => $quoteCurrency,
+                'is_free_shipping' => (bool)$this->quote->getShippingAddress()->getFreeShipping(),
+                'amount_gross'     => $couponsIncludeTax ? $discountAmount : null,
+                'amount_net'       => $couponsIncludeTax ? null : $discountAmount,
+
+            ];
+            $coupons[]                   = $this->couponHelper->dataToEntity($couponData);
+            $returnInvalidCartRuleCoupon = false;
```

Comment:
just one last thing, can you make this variable name shorter? It is just way too long. Smth like `$isCouponValid`